sdk/authentication.py

```diff
@@ -0,0 +1,121 @@
+
+    def __call__(self, r: PreparedRequest) -> PreparedRequest:
+        r.headers["Authorization"] = f"Bearer {self.token_provider.get_token()}"
+        return r
+
+
+class TokenProviderError(Exception):
+    pass
+
+
+class TokenProvider(ABC):
+    __token_cache: Optional[tuple[datetime, str]] = None
+    expiry_window: timedelta = timedelta(minutes=1.0)
+
+    @abstractmethod
+    def _query_token(self) -> dict[str, Any]:
+        raise NotImplementedError
+
+    def get_token(self) -> str:
+        if self.__token_cache:
+            expiry, token = self.__token_cache
+            if expiry - self.expiry_window > datetime.now(tz=timezone.utc):
+                return token
+
+        self.__token_cache = self._refresh_token_cache()
+        return self.__token_cache[1]
+
+    def _refresh_token_cache(self) -> tuple[datetime, str]:
+        try:
+            token_response = self._query_token()
+            expiry = self._extract_expiry(token_response)
+
+            if expiry is None:
+                expiry = datetime.now(tz=timezone.utc)
+
+            return (expiry, token_response["access_token"])
+        except HTTPError as err:
+            raise TokenProviderError(err)
+
+    @staticmethod
+    def _extract_expiry(token_response: dict[str, Any]) -> datetime | None:
+        """Extracts the expiry datetime from the token response.
+
+        Assumes that the actual token provided is correct, but might be in an
+        unexpected format
+
+        Args:
+            token_response: A valid token response dictionary.
+
+        Returns:
+            The time the token will expire, or None if it can't be calculated
+        """
+        expires_in = token_response.get("expires_in", None)
+        if expires_in:
+            return datetime.now(tz=timezone.utc) + timedelta(seconds=float(expires_in))
+
+        try:
+            # We assume the token is valid, but might not be in an expected format
+            token = token_response.get("access_token")
+
+            if (isinstance(token, str)and "." in token):
+                token_timestamp = decode(
+                        token,
+                        algorithms=["HS256"],
+                        options={"verify_signature": False},
+                    ).get("exp")
+                if token_timestamp:
+                    return datetime.fromtimestamp(token_timestamp, tz=timezone.utc)
+
+        except DecodeError:
+            # The token is not in a format that could be parsed as a JWT
+            pass
+        return None
+
+
+class Auth0TokenProvider(TokenProvider):
+    def __init__(self, username: str, password: str):
+        self.username = username
+        self.password = password
+
+        #   This is public, so we can put it in the code here.
+        self.public_client_id = "5QtfSu1UV118Iz6By6IJRSNoDrLbAiOv"
+        self.audience = "https://apis.dev.pasqal.cloud/account/api/v1"
+        self.realm = "pcs-users"
+        self.domain = "pasqal-dev.eu.auth0.com"
```

Comment:
I'm not sure this should be pointing to the dev environment by default.

Reply:
You might be right, I didn't even notice the `-dev` tbh. Nice catch